Fix the decoding issues

bobqianic opened the pull request.

HaujetZhao commented:
I wonder why the CJK token in json output by openai/whisper don't split. Is it because openai/whisper did some afterprocessing of the tokens?

bobqianic commented:
> I wonder why the CJK token in json output by openai/whisper don't split. Is it because openai/whisper did some afterprocessing of the tokens?

Yes. See [this ](https://github.com/openai/whisper/blob/ba3f3cd54b0e5b8ce1ab3de13e32122d0d5f98ab/whisper/tokenizer.py#L286-L309)

bobqianic commented:
I'm really busy this week, but I should have some free time next week. I can probably finish and merge this PR into the master branch by the end of next week.

AeneasZhu commented:
> I'm really busy this week, but I should have some free time next week. I can probably finish and merge this PR into the master branch by the end of next week.

Could this pull fix the hallucination or repetion produced by large-v3?

bobqianic commented:
> Could this pull fix the hallucination or repetion produced by large-v3?

While this `Heuristic hallucination detection` might reduce hallucinations and repetitions to some extent, it's unlikely to eliminate them entirely when using large-v3.

AeneasZhu commented:
> > Could this pull fix the hallucination or repetion produced by large-v3?
> 
> While this `Heuristic hallucination detection` might reduce hallucinations and repetitions to some extent, it's unlikely to eliminate them entirely when using large-v3.

Do you have any idea why large-v3 prouduces much more hallucination than its previous versions? And could there be any solution to this problem by fine-tuning parameters like `beam`, `mc`?

bobqianic commented:
Due to the need to integrate [`zlib`](https://github.com/madler/zlib) for calculating the `compression ratio`, `Use Compression Ratio Instead of Entropy` and `Heuristic Hallucination Detection` will be addressed in separate PRs. This approach will allow for more thorough discussions.
See https://github.com/ggerganov/whisper.cpp/discussions/1461#discussioncomment-8340428

AeneasZhu commented:
> Due to the need to integrate [`zlib`](https://github.com/madler/zlib) for calculating the `compression ratio`, `Use Compression Ratio Instead of Entropy` and `Heuristic Hallucination Detection` will be addressed in separate PRs. This approach will allow for more thorough discussions. See [#1461 (reply in thread)](https://github.com/ggerganov/whisper.cpp/discussions/1461#discussioncomment-8340428)

Is this PR almost finished? I can't wait to see the outcome. By the way, shall I just download whisper.cpp from the master when this  PR is finished. Will you merge it into master?

bobqianic commented:
### Test results

<details>

<summary>Tips for collapsed sections</summary>

### Before:
Filename: `01-03（轻松学中文+第二版+课本2）.wav`  Language: `zh`
```console
C:\Users\qianp\Downloads\whisper-cublas-11.8.0-bin-x64>C:\Users\qianp\Downloads\whisper-cublas-11.8.0-bin-x64\main.exe -m C:\Users\qianp\Downloads\whisper.cpp_build-fix\bin\Release\ggml-model-whisper-large.bin -f C:\Users\qianp\Downloads\whisper.cpp_build-fix\bin\Release\01-03（轻松学中文+第二版+课本2）.wav -l auto
whisper_init_from_file_with_params_no_state: loading model from 'C:\Users\qianp\Downloads\whisper.cpp_build-fix\bin\Release\ggml-model-whisper-large.bin'
whisper_model_load: loading model
whisper_model_load: n_vocab       = 51865
whisper_model_load: n_audio_ctx   = 1500
whisper_model_load: n_audio_state = 1280
whisper_model_load: n_audio_head  = 20
whisper_model_load: n_audio_layer = 32
whisper_model_load: n_text_ctx    = 448
whisper_model_load: n_text_state  = 1280
whisper_model_load: n_text_head   = 20
whisper_model_load: n_text_layer  = 32
whisper_model_load: n_mels        = 80
whisper_model_load: ftype         = 1
whisper_model_load: qntvr         = 0
whisper_model_load: type          = 5 (large)
whisper_model_load: adding 1608 extra tokens
whisper_model_load: n_langs       = 99
ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 3060 Laptop GPU, compute capability 8.6, VMM: yes
whisper_backend_init: using CUDA backend
whisper_model_load:    CUDA0 total size =  3094.49 MB (3 buffers)
whisper_model_load: model size    = 3093.99 MB
whisper_backend_init: using CUDA backend
whisper_init_state: kv self size  =  220.20 MB
whisper_init_state: kv cross size =  245.76 MB
whisper_init_state: compute buffer (conv)   =   33.91 MB
whisper_init_state: compute buffer (encode) =  233.50 MB
whisper_init_state: compute buffer (cross)  =   10.15 MB
whisper_init_state: compute buffer (decode) =  108.99 MB
error: failed to open 'C:\Users\qianp\Downloads\whisper.cpp_build-fix\bin\Release\01-03(?????+???+??2).wav' as WAV file
error: failed to read WAV file 'C:\Users\qianp\Downloads\whisper.cpp_build-fix\bin\Release\01-03(?????+???+??2).wav'

whisper_print_timings:     load time =  6602.47 ms
whisper_print_timings:     fallbacks =   0 p /   0 h
whisper_print_timings:      mel time =     0.00 ms
whisper_print_timings:   sample time =     0.00 ms /     1 runs (    0.00 ms per run)
whisper_print_timings:   encode time =     0.00 ms /     1 runs (    0.00 ms per run)
whisper_print_timings:   decode time =     0.00 ms /     1 runs (    0.00 ms per run)
whisper_print_timings:   batchd time =     0.00 ms /     1 runs (    0.00 ms per run)
whisper_print_timings:   prompt time =     0.00 ms /     1 runs (    0.00 ms per run)
whisper_print_timings:    total time =  6608.45 ms
```

Filename: `chinese.wav`  Language: `zh`
```console
C:\Users\qianp\Downloads\whisper-cublas-11.8.0-bin-x64>C:\Users\qianp\Downloads\whisper-cublas-11.8.0-bin-x64\main.exe -m C:\Users\qianp\Downloads\whisper.cpp_build-fix\bin\Release\ggml-model-whisper-large.bin -f C:\Users\qianp\Downloads\whisper.cpp_build-fix\bin\Release\chinese.wav -l auto
whisper_init_from_file_with_params_no_state: loading model from 'C:\Users\qianp\Downloads\whisper.cpp_build-fix\bin\Release\ggml-model-whisper-large.bin'
whisper_model_load: loading model
whisper_model_load: n_vocab       = 51865
whisper_model_load: n_audio_ctx   = 1500
whisper_model_load: n_audio_state = 1280
whisper_model_load: n_audio_head  = 20
whisper_model_load: n_audio_layer = 32
whisper_model_load: n_text_ctx    = 448
whisper_model_load: n_text_state  = 1280
whisper_model_load: n_text_head   = 20
whisper_model_load: n_text_layer  = 32
whisper_model_load: n_mels        = 80
whisper_model_load: ftype         = 1
whisper_model_load: qntvr         = 0
whisper_model_load: type          = 5 (large)
whisper_model_load: adding 1608 extra tokens
whisper_model_load: n_langs       = 99
ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 3060 Laptop GPU, compute capability 8.6, VMM: yes
whisper_backend_init: using CUDA backend
whisper_model_load:    CUDA0 total size =  3094.49 MB (3 buffers)
whisper_model_load: model size    = 3093.99 MB
whisper_backend_init: using CUDA backend
whisper_init_state: kv self size  =  220.20 MB
whisper_init_state: kv cross size =  245.76 MB
whisper_init_state: compute buffer (conv)   =   33.91 MB
whisper_init_state: compute buffer (encode) =  233.50 MB
whisper_init_state: compute buffer (cross)  =   10.15 MB
whisper_init_state: compute buffer (decode) =  108.99 MB

system_info: n_threads = 4 / 20 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | METAL = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 0 | VSX = 0 | CUDA = 1 | COREML = 0 | OPENVINO = 0 |

main: processing 'C:\Users\qianp\Downloads\whisper.cpp_build-fix\bin\Release\chinese.wav' (713721 samples, 44.6 sec), 4 threads, 1 processors, 5 beams + best of 5, lang = auto, task = transcribe, timestamps = 1 ...

whisper_full_with_state: auto-detected language: zh (p = 0.994848)

[00:00:00.000 --> 00:00:04.800]  Σ╜áτê╖τê╖σÑ╢σÑ╢Σ╜Åσ£¿σô¬σä┐?
[00:00:04.800 --> 00:00:07.800]  Σ╗ûΣ╗¼Σ╜Åσ£¿σìùΣ║¼πÇé
[00:00:07.800 --> 00:00:12.800]  Σ╗ûΣ╗¼Φ╖ƒµêæσÅöσÅöσÆîσ⌐╢σ⌐╢Σ╕ÇΦ╡╖Σ╜ÅπÇé
[00:00:12.800 --> 00:00:16.800]  Σ╜áσÅöσÅöµÿ»σô¬Σ╕Çσ╣┤τ╗ôσ⌐ÜτÜä?
[00:00:16.800 --> 00:00:19.800]  Σ╗ûµÿ»σÄ╗σ╣┤τ╗ôσ⌐ÜτÜäπÇé
[00:00:19.800 --> 00:00:23.800]  Σ╜áσÅöσÅöσüÜΣ╗ÇΣ╣êσ╖ÑΣ╜£?
[00:00:23.800 --> 00:00:25.800]  σ£¿σô¬σä┐σ╖ÑΣ╜£?
[00:00:25.800 --> 00:00:27.800]  µêæσÅöσÅöµÿ»ΦÇüσ╕êπÇé
[00:00:28.000 --> 00:00:31.800]  Σ╗ûσ£¿σÉîσÉìΣ╕¡σ¡ªσ╖ÑΣ╜£πÇé
[00:00:31.800 --> 00:00:37.800]  Σ╜áσ╕╕σ╕╕Φ╖ƒΣ╜áτê╕τê╕σ«╢τÜäΣ║▓µêÜΦºüΘ¥óσÉù?
[00:00:37.800 --> 00:00:40.800]  µêæσ╕╕σ╕╕Φ╖ƒΣ╗ûΣ╗¼ΦºüΘ¥óπÇé
[00:00:40.800 --> 00:00:42.800]  ΦªïΘ¥ó
[00:00:42.800 --> 00:00:44.800]  µêæσÅ¬µâ│τƒÑΘüôΣ╜áσê░σ║òµÿ»Φ░üσòè?


whisper_print_timings:     load time =  4116.39 ms
whisper_print_timings:     fallbacks =   1 p /   0 h
whisper_print_timings:      mel time =    34.72 ms
whisper_print_timings:   sample time =   334.78 ms /   744 runs (    0.45 ms per run)
whisper_print_timings:   encode time =  1144.36 ms /     5 runs (  228.87 ms per run)
whisper_print_timings:   decode time =   226.45 ms /     5 runs (   45.29 ms per run)
whisper_print_timings:   batchd time =  3613.09 ms /   727 runs (    4.97 ms per run)
whisper_print_timings:   prompt time =   186.71 ms /    88 runs (    2.12 ms per run)
whisper_print_timings:    total time =  9686.97 ms
```

Filename: `chinese.wav`  Language: `zh` + `chcp 65001`
```console
C:\Users\qianp\Downloads\whisper-cublas-11.8.0-bin-x64>C:\Users\qianp\Downloads\whisper-cublas-11.8.0-bin-x64\main.exe -m C:\Users\qianp\Downloads\whisper.cpp_build-fix\bin\Release\ggml-model-whisper-large.bin -f C:\Users\qianp\Downloads\whisper.cpp_build-fix\bin\Release\chinese.wav -l auto
whisper_init_from_file_with_params_no_state: loading model from 'C:\Users\qianp\Downloads\whisper.cpp_build-fix\bin\Release\ggml-model-whisper-large.bin'
whisper_model_load: loading model
whisper_model_load: n_vocab       = 51865
whisper_model_load: n_audio_ctx   = 1500
whisper_model_load: n_audio_state = 1280
whisper_model_load: n_audio_head  = 20
whisper_model_load: n_audio_layer = 32
whisper_model_load: n_text_ctx    = 448
whisper_model_load: n_text_state  = 1280
whisper_model_load: n_text_head   = 20
whisper_model_load: n_text_layer  = 32
whisper_model_load: n_mels        = 80
whisper_model_load: ftype         = 1
whisper_model_load: qntvr         = 0
whisper_model_load: type          = 5 (large)
whisper_model_load: adding 1608 extra tokens
whisper_model_load: n_langs       = 99
ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 3060 Laptop GPU, compute capability 8.6, VMM: yes
whisper_backend_init: using CUDA backend
whisper_model_load:    CUDA0 total size =  3094.49 MB (3 buffers)
whisper_model_load: model size    = 3093.99 MB
whisper_backend_init: using CUDA backend
whisper_init_state: kv self size  =  220.20 MB
whisper_init_state: kv cross size =  245.76 MB
whisper_init_state: compute buffer (conv)   =   33.91 MB
whisper_init_state: compute buffer (encode) =  233.50 MB
whisper_init_state: compute buffer (cross)  =   10.15 MB
whisper_init_state: compute buffer (decode) =  108.99 MB

system_info: n_threads = 4 / 20 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | METAL = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 0 | VSX = 0 | CUDA = 1 | COREML = 0 | OPENVINO = 0 |

main: processing 'C:\Users\qianp\Downloads\whisper.cpp_build-fix\bin\Release\chinese.wav' (713721 samples, 44.6 sec), 4 threads, 1 processors, 5 beams + best of 5, lang = auto, task = transcribe, timestamps = 1 ...

whisper_full_with_state: auto-detected language: zh (p = 0.994848)

[00:00:00.000 --> 00:00:04.800]  你爷爷奶奶住在哪儿?
[00:00:04.800 --> 00:00:07.800]  他们住在南京。
[00:00:07.800 --> 00:00:12.800]  他们跟我叔叔和婶婶一起住。
[00:00:12.800 --> 00:00:16.800]  你叔叔是哪一年结婚的?
[00:00:16.800 --> 00:00:19.800]  他是去年结婚的。
[00:00:19.800 --> 00:00:23.800]  你叔叔做什么工作?
[00:00:23.800 --> 00:00:25.800]  在哪儿工作?
[00:00:25.800 --> 00:00:27.800]  我叔叔是老师。
[00:00:28.000 --> 00:00:31.800]  他在同名中学工作。
[00:00:31.800 --> 00:00:37.800]  你常常跟你爸爸家的亲戚见面吗?
[00:00:37.800 --> 00:00:40.800]  我常常跟他们见面。
[00:00:40.800 --> 00:00:42.800]  見面
[00:00:42.800 --> 00:00:44.800]  我只想知道你到底是谁啊?


whisper_print_timings:     load time =  4031.71 ms
whisper_print_timings:     fallbacks =   1 p /   0 h
whisper_print_timings:      mel time =    31.19 ms
whisper_print_timings:   sample time =   355.73 ms /   744 runs (    0.48 ms per run)
whisper_print_timings:   encode time =  1148.79 ms /     5 runs (  229.76 ms per run)
whisper_print_timings:   decode time =   227.00 ms /     5 runs (   45.40 ms per run)
whisper_print_timings:   batchd time =  3931.32 ms /   727 runs (    5.41 ms per run)
whisper_print_timings:   prompt time =   190.18 ms /    88 runs (    2.16 ms per run)
whisper_print_timings:    total time =  9930.92 ms
```

### After:

Filename: `01-03（轻松学中文+第二版+课本2）.wav`  Language: `zh`
```console
C:\Users\qianp\Downloads\whisper.cpp_build-fix\bin\Release>C:\Users\qianp\Downloads\whisper.cpp_build-fix\bin\Release\main.exe -m C:\Users\qianp\Downloads\whisper.cpp_build-fix\bin\Release\ggml-model-whisper-large.bin -f C:\Users\qianp\Downloads\whisper.cpp_build-fix\bin\Release\01-03（轻松学中文+第二版+课本2）.wav -l auto
whisper_init_from_file_with_params_no_state: loading model from 'C:\Users\qianp\Downloads\whisper.cpp_build-fix\bin\Release\ggml-model-whisper-large.bin'
whisper_model_load: loading model
whisper_model_load: n_vocab       = 51865
whisper_model_load: n_audio_ctx   = 1500
whisper_model_load: n_audio_state = 1280
whisper_model_load: n_audio_head  = 20
whisper_model_load: n_audio_layer = 32
whisper_model_load: n_text_ctx    = 448
whisper_model_load: n_text_state  = 1280
whisper_model_load: n_text_head   = 20
whisper_model_load: n_text_layer  = 32
whisper_model_load: n_mels        = 80
whisper_model_load: ftype         = 1
whisper_model_load: qntvr         = 0
whisper_model_load: type          = 5 (large)
whisper_model_load: adding 1608 extra tokens
whisper_model_load: n_langs       = 99
ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 3060 Laptop GPU, compute capability 8.6, VMM: yes
whisper_backend_init: using CUDA backend
whisper_model_load:    CUDA0 total size =  3094.49 MB (3 buffers)
whisper_model_load: model size    = 3093.99 MB
whisper_backend_init: using CUDA backend
whisper_init_state: kv self size  =  220.20 MB
whisper_init_state: kv cross size =  245.76 MB
whisper_init_state: compute buffer (conv)   =   33.91 MB
whisper_init_state: compute buffer (encode) =  233.50 MB
whisper_init_state: compute buffer (cross)  =   10.15 MB
whisper_init_state: compute buffer (decode) =  108.99 MB

system_info: n_threads = 4 / 20 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | METAL = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 0 | VSX = 0 | CUDA = 1 | COREML = 0 | OPENVINO = 0 |

run: processing 'C:\Users\qianp\Downloads\whisper.cpp_build-fix\bin\Release\01-03（轻松学中文+第二版+课本2）.wav' (713721 samples, 44.6 sec), 4 threads, 1 processors, 5 beams + best of 5, lang = auto, task = transcribe, timestamps = 1 ...

whisper_full_with_state: auto-detected language: zh (p = 0.994848)

[00:00:01.000 --> 00:00:05.000]  你爷爷奶奶住在哪儿?
[00:00:05.000 --> 00:00:08.000]  他们住在南京。
[00:00:08.000 --> 00:00:13.000]  他们跟我叔叔和婶婶一起住。
[00:00:13.000 --> 00:00:17.000]  你叔叔是哪一年结婚的?
[00:00:17.000 --> 00:00:20.000]  他是去年结婚的。
[00:00:20.000 --> 00:00:24.000]  你叔叔做什么工作?
[00:00:24.000 --> 00:00:26.000]  在哪儿工作?
[00:00:26.000 --> 00:00:28.000]  我叔叔是老师。
[00:00:28.000 --> 00:00:32.000]  他在同名中学工作。
[00:00:32.000 --> 00:00:38.000]  你常常跟你爸爸家的亲戚见面吗?
[00:00:38.000 --> 00:00:41.000]  我常常跟他们见面。


whisper_print_timings:     load time =  4086.30 ms
whisper_print_timings:     fallbacks =   0 p /   0 h
whisper_print_timings:      mel time =    38.71 ms
whisper_print_timings:   sample time =   284.85 ms /   608 runs (    0.47 ms per run)
whisper_print_timings:   encode time =   692.73 ms /     3 runs (  230.91 ms per run)
whisper_print_timings:   decode time =   167.72 ms /     1 runs (  167.72 ms per run)
whisper_print_timings:   batchd time =  2644.88 ms /   603 runs (    4.39 ms per run)
whisper_print_timings:   prompt time =   195.07 ms /    89 runs (    2.19 ms per run)
whisper_print_timings:    total time =  8294.52 ms
```

</details>

ggerganov commented on examples/console.cpp:
Should remove from `console.cpp` what is not used

Even better - the UTF8 / UTF16 change should go into separate PR. This is a big change that can affect significantly the quality of `whisper.cpp`, so the PR should contain the minimal set of changes necessary for that

ggerganov commented on whisper.cpp:
```diff
@@ -2936,53 +3119,23 @@ static bool log_mel_spectrogram(
-        }
-    }
+    // this enables parallel processing
+    auto words = bpe_gpt2_preprocess(" " + text);
```
Are you sure that whitespace should be here?

bobqianic replied:
Yes. See [transcribe.py L230-L234](https://github.com/openai/whisper/blob/ba3f3cd54b0e5b8ce1ab3de13e32122d0d5f98ab/whisper/transcribe.py#L230-L234)

ggerganov commented on whisper.cpp:
```diff
@@ -4702,17 +4963,17 @@ static void whisper_process_logits(
             // logsumexp over timestamps
+            // ref: https://pytorch.org/docs/stable/generated/torch.logsumexp.html
             float timestamp_logprob = -INFINITY;
             {
                 float logsumexp = 0.0f;
-                const float logprob_max = *std::max_element(logprobs.begin() + vocab.token_beg, logprobs.end());
                 for (int i = vocab.token_beg; i < n_logits; ++i) {
                     if (logprobs[i] > -INFINITY) {
-                        logsumexp += expf(logprobs[i] - logprob_max);
+                        logsumexp += expf(logprobs[i]);
                     }
                 }
                 if (logsumexp > 0.0f) {
-                    timestamp_logprob = logf(logsumexp) + logprob_max;
+                    timestamp_logprob = logf(logsumexp);
                 }
```
This change is not necessary - the old implementation is a more numerically stable way of computing the `logsumexp`

bobqianic replied:
OK, I will revert that.

ggerganov commented:
Does this change have any positive effect for `whisper-v3` or it's still repeating stuff?

bobqianic commented:
> Does this change have any positive effect for `whisper-v3` or it's still repeating stuff?

Overall, there is a positive effect on all models. However, given the limitations of Whisper v3, any improvements made to it still fall short of expectations. See the analysis provided by Deepgram: [Whisper v3 Results](https://deepgram.com/learn/whisper-v3-results).

bobqianic commented:
I think that to completely avoid hallucination, the best approach is similar to using DTW to calculate token timestamps. By comparing these with cross-attention weights, we can definitely identify anomalies if there are any hallucinations.

felrock reviewed (commented):
looks good to me, but I think @ggerganov needs to approve it

bobqianic commented:
> Did you run some tests?

I've done some initial testing, and the results are promising. However, I need a bit more time to conduct a comprehensive analysis. You can already notice the difference by testing a few audio files. Currently, I'm downloading the Common Voice Corpus 15.0, which is over 100GB, so completing the testing will take a little while.

jettoblack commented:
@bobqianic I'm very appreciative of this work and very excited to see this branch implemented, but getting some bad results with weird non-speech tokens at the beginning of many files, this problem does not happen in master branch.

## Example 1:
wav file: https://www.dropbox.com/scl/fi/bdz7lx4khunq3kiauyus8/shermer.wav?rlkey=hzy02rkewjb4pwoamp9whch4b&dl=0

### Command:
`./main -m ggml-largev2.bin -f shermer.wav`

### Output of master branch @ 434b8f3b (current):
[00:00:00.000 --> 00:00:09.000]   [music]
[00:00:09.000 --> 00:00:12.000]   [applause]
[00:00:12.000 --> 00:00:14.000]   Hey, I am Michael Shermer, the director of the Skeptic Society,
...

### Output of this PR @ c0277e3:
[00:00:00.000 --> 00:00:07.000]   Transcriber's Name Reviewer's Name
[00:00:12.340 --> 00:00:14.300]   I am Michael Shermer, the director of the Skeptic Society,
...

## Example 2 with translate fr to en:
wav file: https://www.dropbox.com/scl/fi/1go0yxkr10vwhfyxs76vz/french.wav?rlkey=312gc5qmw3r31ovh003410hyb&dl=0

### Command:
`./main -m ggml-largev2.bin -f french.wav -l fr -tr`

### Output of master branch @ 434b8f3b (current):
[00:00:00.000 --> 00:00:04.000]   (Music)
[00:00:04.000 --> 00:00:07.000]   (Applause)
[00:00:07.000 --> 00:00:20.000]   I am a champion of France.
...

### Output of this PR @ c0277e3:
[00:00:00.000 --> 00:00:17.000]   Translation & subtitling by Quentin Dewaghe Traduction & sous-titrage par Quentin Dewaghe q.dewaghe.com
[00:00:17.000 --> 00:00:20.000]   I'm a champion of France.
...


Any idea why these non-speech tokens like "Transcriber's Name Reviewer's Name" are being output as speech at the beginning? Thanks again.

bobqianic commented:
> Any idea why these non-speech tokens like "Transcriber's Name Reviewer's Name" are being output as speech at the beginning? Thanks again.

Thank you for letting me know. It seems the primary issue stems from my having suppressed non-speech tokens, which has resulted in symbols like `(` and `[` having a zero probability of appearing. While this approach enhances the overall quality, it clearly didn't account for situations like yours, which I hadn't anticipated. As mentioned, I'll conduct further tests and explore ways to address this issue.

bobqianic commented:
@jettoblack I've added a heuristic for detecting repetitive hallucinations, which you can disable via parameters if you prefer. Additionally, I've removed the tokens `(` `)` `[` and `]` from the list of tokens to be suppressed, so they will remain unaffected even when suppression mode is enabled. 

### Output of this PR @ https://github.com/ggerganov/whisper.cpp/pull/1768/commits/476dff454488af4710d4ad5f230fb8a7ac810d6f:

[00:00:00.000 --> 00:00:17.000]   [Music]
[00:00:17.000 --> 00:00:20.000]   I am a champion of France.
...

jettoblack commented:
@bobqianic The repetition heuristic seems to be working well so far. I'm seeing fewer hallucinations on silent intervals. I looked at the code and this is unrelated to the non-speech token changes, right?

I'm not so sure about the non-speech token changes. With your latest commit I see fewer cases of the problem I mentioned above, but it's still happening a lot. One example I got just now in the sg1.wav file I sent you previously on Discord:

[00:57:11.700 --> 00:57:14.700]   (c) 2014 University of Georgia College of Agricultural and Environmental Sciences UGA Extension Office of Communications and Creative Services

A hallucination like ♪♪ or repeated text is far less objectionable than someone else's copyright notice or translator's notes which is what I'm getting a lot of.

This change also removes many useful tokens from the output, like quotation marks and music notes.  Using the -nsnst option restores these tokens but that causes this issue to be much worse, and I've caught a lot more cases of it occurring in many files, including in the middle of files not just the beginning. If these were the only two options I'd leave suppression enabled, but master branch includes these useful tokens without this hallucination problem. 

It might be helpful to compare the output of a branch with the other fixes of this PR excluding the non-speech token changes, or at least have a way to turn those completely off and go back to master branch behavior.

bobqianic commented:
> I looked at the code and this is unrelated to the non-speech token changes, right?

Yes. In situations where the model exhibits hallucinations with high confidence (avg_log_probs), this `non-speech token` approach will not be effective. The heuristic repetition check that I've implemented serves as a workaround for the compression ratio check. Implementing compression in C++ can be challenging without using third-party libraries. In the official implementation by OpenAI, both the `compression ratio` and `non-speech tokens` anti-hallucination mechanisms are utilized.

> [00:57:11.700 --> 00:57:14.700] (c) 2014 University of Georgia College of Agricultural and Environmental Sciences UGA Extension Office of Communications and Creative Services

Which branch are you using? I can't find the hallucinations you mentioned.

![image](https://github.com/ggerganov/whisper.cpp/assets/129547291/e4076505-b25e-42da-a528-120d44b1914e)

jettoblack commented:
> Which branch are you using? I can't find the hallucinations you mentioned.

I was using this PR @ 476dff4, unless I did something wrong, but this was on a Mac using the Metal gpu backend so that could make a difference. I'll retest on CPU and CUDA shortly and let you know.

ukolovda commented:
Hi!  

@bobqianic  new version is very robust!

On my test files, main branch emit 10 hallucinations on 26 WAV files (model `ggml-large-v2`, russian language). 
With this PR it give only 2 hallucination. It is very fine result!!!

But `example/server` doesn't work at all, both CPU and CUDA versions. It returns empty text without any errors. 
I try patch and append parameters (heuristic and other), but it not help. With `--print-progress` it print progress, but not result text.

Also it give the error on specific file:
`500 Internal Server Error
map::at`

What can we do for fix it, how do you think?

Run server command:
```bash
/usr/src/whisper.cpp-bobqianic/server -m ../../models/ggml-large-v2.bin -l ru --print-progress --print-realtime -nt -nf

whisper_init_from_file_with_params_no_state: loading model from '../../models/ggml-large-v2.bin'
whisper_model_load: loading model
whisper_model_load: n_vocab       = 51865
whisper_model_load: n_audio_ctx   = 1500
whisper_model_load: n_audio_state = 1280
whisper_model_load: n_audio_head  = 20
whisper_model_load: n_audio_layer = 32
whisper_model_load: n_text_ctx    = 448
whisper_model_load: n_text_state  = 1280
whisper_model_load: n_text_head   = 20
whisper_model_load: n_text_layer  = 32
whisper_model_load: n_mels        = 80
whisper_model_load: ftype         = 1
whisper_model_load: qntvr         = 0
whisper_model_load: type          = 5 (large)
whisper_model_load: adding 1608 extra tokens
whisper_model_load: n_langs       = 99
ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 4070, compute capability 8.9, VMM: yes
whisper_backend_init: using CUDA backend
whisper_model_load:    CUDA0 total size =  3094.49 MB (3 buffers)
whisper_model_load: model size    = 3093.99 MB
whisper_backend_init: using CUDA backend
whisper_init_state: kv self size  =  220.20 MB
whisper_init_state: kv cross size =  245.76 MB
whisper_init_state: compute buffer (conv)   =   33.91 MB
whisper_init_state: compute buffer (encode) =  233.50 MB
whisper_init_state: compute buffer (cross)  =   10.15 MB
whisper_init_state: compute buffer (decode) =  108.99 MB

whisper server listening at http://127.0.0.1:8080

Received request: 0f3657ce-6352-4cbb-a88f-b39dc6a37a34-1-14.wav
Successfully loaded 0f3657ce-6352-4cbb-a88f-b39dc6a37a34-1-14.wav

system_info: n_threads = 4 / 12 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | METAL = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | CUDA = 1 | COREML = 0 | OPENVINO = 0 | 

operator(): processing '0f3657ce-6352-4cbb-a88f-b39dc6a37a34-1-14.wav' (168960 samples, 10.6 sec), 4 threads, 1 processors, lang = ru, task = transcribe, timestamps = 0 ...

Running whisper.cpp inference on 0f3657ce-6352-4cbb-a88f-b39dc6a37a34-1-14.wav
Received request: 0f3657ce-6352-4cbb-a88f-b39dc6a37a34-1-15.wav
Successfully loaded 0f3657ce-6352-4cbb-a88f-b39dc6a37a34-1-15.wav

system_info: n_threads = 4 / 12 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | METAL = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | CUDA = 1 | COREML = 0 | OPENVINO = 0 | 

operator(): processing '0f3657ce-6352-4cbb-a88f-b39dc6a37a34-1-15.wav' (235200 samples, 14.7 sec), 4 threads, 1 processors, lang = ru, task = transcribe, timestamps = 0 ...

Running whisper.cpp inference on 0f3657ce-6352-4cbb-a88f-b39dc6a37a34-1-15.wav

whisper_print_progress_callback: progress = 204%
Received request: 0f3657ce-6352-4cbb-a88f-b39dc6a37a34-1-16.wav
Successfully loaded 0f3657ce-6352-4cbb-a88f-b39dc6a37a34-1-16.wav

system_info: n_threads = 4 / 12 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | METAL = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | CUDA = 1 | COREML = 0 | OPENVINO = 0 | 

operator(): processing '0f3657ce-6352-4cbb-a88f-b39dc6a37a34-1-16.wav' (512000 samples, 32.0 sec), 4 threads, 1 processors, lang = ru, task = transcribe, timestamps = 0 ...

Running whisper.cpp inference on 0f3657ce-6352-4cbb-a88f-b39dc6a37a34-1-16.wav

whisper_print_progress_callback: progress =  93%

whisper_print_progress_callback: progress = 187%
Received request: 0f3657ce-6352-4cbb-a88f-b39dc6a37a34-1-18.wav
Successfully loaded 0f3657ce-6352-4cbb-a88f-b39dc6a37a34-1-18.wav

system_info: n_threads = 4 / 12 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | METAL = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | CUDA = 1 | COREML = 0 | OPENVINO = 0 | 

operator(): processing '0f3657ce-6352-4cbb-a88f-b39dc6a37a34-1-18.wav' (115520 samples, 7.2 sec), 4 threads, 1 processors, lang = ru, task = transcribe, timestamps = 0 ...

Running whisper.cpp inference on 0f3657ce-6352-4cbb-a88f-b39dc6a37a34-1-18.wav

whisper_print_progress_callback: progress = 416%
...
```

Send file command:
```bash
curl localhost:8080/inference -H "Content-Type: multipart/form-data" -F file="@${filename}"
```


Thank you!

felrock commented:
Hello @ukolovda I took a look at this yesterday evening. Whats missing in server.cpp is what you mentioned:
* heuristics
* supress_nst
* no_speech_thold

I got an output in the terminal by circumventing the **print_realtime** flag(instead of using a callback segment). So the model does in fact generate the output string but for some unknown reason `whisper_full_n_segments(ctx)` returns 0. I try to check this a bit more tomorrow.

ukolovda commented:
> I got an output in the terminal by circumventing the **print_realtime** flag(instead of using a callback segment). So the model does in fact generate the output string but for some unknown reason `whisper_full_n_segments(ctx)` returns 0. 

Hello, @felrock !

Thank you!

ukolovda commented:
Append issue with zero-padded WAV.
https://github.com/ggerganov/whisper.cpp/issues/1881

ukolovda commented:
File from #1881 (zero filled WAV) give a gallucination in this version also.

```bash
$ ../whisper.cpp-bobqianic/main -m ./models/ggml-large-v3.bin -l ru --threads 8 -mc 0 samples/zeroes.wav
whisper_init_from_file_with_params_no_state: loading model from './models/ggml-large-v3.bin'
whisper_model_load: loading model
whisper_model_load: n_vocab       = 51866
whisper_model_load: n_audio_ctx   = 1500
whisper_model_load: n_audio_state = 1280
whisper_model_load: n_audio_head  = 20
whisper_model_load: n_audio_layer = 32
whisper_model_load: n_text_ctx    = 448
whisper_model_load: n_text_state  = 1280
whisper_model_load: n_text_head   = 20
whisper_model_load: n_text_layer  = 32
whisper_model_load: n_mels        = 128
whisper_model_load: ftype         = 1
whisper_model_load: qntvr         = 0
whisper_model_load: type          = 5 (large v3)
whisper_model_load: adding 1609 extra tokens
whisper_model_load: n_langs       = 100
ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 4070, compute capability 8.9, VMM: yes
whisper_backend_init: using CUDA backend
whisper_model_load:    CUDA0 total size =  3094,86 MB (3 buffers)
whisper_model_load: model size    = 3094,36 MB
whisper_backend_init: using CUDA backend
whisper_init_state: kv self size  =  220,20 MB
whisper_init_state: kv cross size =  245,76 MB
whisper_init_state: compute buffer (conv)   =   35,50 MB
whisper_init_state: compute buffer (encode) =  233,50 MB
whisper_init_state: compute buffer (cross)  =   10,15 MB
whisper_init_state: compute buffer (decode) =  108,99 MB

system_info: n_threads = 8 / 12 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | METAL = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | CUDA = 1 | COREML = 0 | OPENVINO = 0 | 

run: processing 'samples/zeroes.wav' (19200 samples, 1,2 sec), 8 threads, 1 processors, 5 beams + best of 5, lang = ru, task = transcribe, timestamps = 1 ...


[00:00:00.000 --> 00:00:29.980]   Продолжение следует...


whisper_print_timings:     load time =   781,61 ms
whisper_print_timings:     fallbacks =   0 p /   0 h
whisper_print_timings:      mel time =     4,81 ms
whisper_print_timings:   sample time =    28,10 ms /    79 runs (    0,36 ms per run)
whisper_print_timings:   encode time =   162,31 ms /     1 runs (  162,31 ms per run)
whisper_print_timings:   decode time =     0,00 ms /     1 runs (    0,00 ms per run)
whisper_print_timings:   batchd time =   482,89 ms /    77 runs (    6,27 ms per run)
whisper_print_timings:   prompt time =     0,00 ms /     1 runs (    0,00 ms per run)
whisper_print_timings:    total time =  1502,74 ms
```

linmi commented:
-output-json-full has problems with the output format.
![CleanShot 2024-02-21 at 16 41 35@2x](https://github.com/ggerganov/whisper.cpp/assets/13565344/da6e4119-7966-4586-882e-514cb700ca42)

thewh1teagle commented:
What's the status of this PR? is it safe to use?
I experience decoding issues
https://github.com/thewh1teagle/vibe/issues/34

jwijffels commented:
I'm thinking about including this pull request in the R wrapper at [audio.whisper](https://github.com/bnosac/audio.whisper) . There the current approach to handle some of the hallucinations is to use R packages [audio.vadwebrtc](https://github.com/bnosac/audio.vadwebrtc) or  [audio.vadsilero](https://github.com/bnosac/audio.vadsilero) to detect silences or general non-voiced signals and either

- instead of looping over different files in the main loop, loop over the detected non-silence sections in the audio. 
- or create a new audio file with only the voiced audio and recompute the timestamps later on by adding what was left out

I haven't looked into the extreme details on this pull request (only skimmed through the logic which was changed in main.cpp and whisper.cpp) but would it make sense already to incorporate this pull request in [audio.whisper](https://github.com/bnosac/audio.whisper) or are there a lot of changes to be expected here?

bobqianic commented:
> @bobqianic are you pursuing this at the moment?

No, at least not in May. I'm really tied up with a lot of things this month.

bygreencn commented:
> I'm thinking about including this pull request in the R wrapper at [audio.whisper](https://github.com/bnosac/audio.whisper) . There the current approach to handle some of the hallucinations is to use R packages [audio.vadwebrtc](https://github.com/bnosac/audio.vadwebrtc) or [audio.vadsilero](https://github.com/bnosac/audio.vadsilero) to detect silences or general non-voiced signals and either
> 
> * instead of looping over different files in the main loop, loop over the detected non-silence sections in the audio.
> * or create a new audio file with only the voiced audio and recompute the timestamps later on by adding what was left out
> 
> I haven't looked into the extreme details on this pull request (only skimmed through the logic which was changed in main.cpp and whisper.cpp) but would it make sense already to incorporate this pull request in [audio.whisper](https://github.com/bnosac/audio.whisper) or are there a lot of changes to be expected here or is this pull request going to be split into a BPE change (#1854) and a change regarding how to handle non-speech?

The best way to include [Silero](https://github.com/snakers4/silero-vad) Voice Activity into whisper.cpp is to add thirdparty package of onnxruntime**1.12.1** dll, then call silero onnx model. My branch had added it. Even VAD, the hallucinations on silent intervals is also happenning.

IntendedConsequence commented:
> The best way to include [Silero](https://github.com/snakers4/silero-vad) Voice Activity into whisper.cpp is to add thirdparty package of onnxruntime**1.12.1** dll, then call silero onnx model. My branch had added it. Even VAD, the hallucinations on silent intervals is also happenning.

I recommend considering a previous Silero VAD version, namely v3.1. The current version v4 (at the moment of writing) often hallucinates speech on lengthy chunks of silent or near-silent audio segments.
https://github.com/snakers4/silero-vad/issues/369
https://github.com/snakers4/silero-vad/issues/396

But you have to add a heavyweight dependency like onnxruntime just to run a 750KB model. The smallest size I could possibly reduce onnxruntime.dll to was about 2.2MB, which is still 3x the size of silero weights, and requires a lengthy custom build of onnxruntime from source with reduced operator set configs and other size reduction options. And prebuilt redistributables are easily 5-9 MB or more.

I have a working Silero v3.1 implementation in pure C, but as much as I would like to suggest it as an option, the code is quite bad, I wrote it as a personal project for learning low level neural nets.

ziegenberg commented:
@bobqianic, Could you rebase your changes? I'd like to test those improvements of yours with production data on our setup.

ziegenberg commented on examples/main/main.cpp:
```diff
@@ -94,7 +97,7 @@ struct whisper_params {
-void whisper_print_usage(int argc, char ** argv, const whisper_params & params);
+void whisper_print_usage(int argc, const char ** argv, const whisper_params & params);
 
 char* whisper_param_turn_lowercase(char* in){
```
I think, this needs a `const`.

```suggestion
char* whisper_param_turn_lowercase(const char* in){
```

bobqianic replied:
I'm still fixing the compatibility issue. A lot of things need to be tested since we added DTW, and some aspects may not be compatible. Give me a few hours.

bobqianic commented:
@ziegenberg I did some testing, and it LGTM. If the CI is mostly green, you can proceed with your testing now.

ziegenberg commented:
I already did some testing and fixed some of the errors on my own. Looks promising. I see less hallucinations, but I need to do some more statistics. I will switch to your branch for the next tests.

Is your PR #1854 also related to this improvement?

bobqianic commented:
> Is your PR #1854 also related to this improvement?
 
 PR #1854 is a subset of this PR, meaning this PR includes everything in PR #1854.

ziegenberg commented:
What data/statistics would you need from my side to consider this PR validated and get it merged?

bobqianic commented:
> What data/statistics would you need from my side to consider this PR validated and get it merged?

Thank you. If you have the ground truth text, please calculate the WER.

Makememo commented:
I tested the output of anime using medium.en and found a problem with time axis recognition in the middle.

file: https://dropover.cloud/f7

<img width="1624" alt="2024-06-27 09 20 40" src="https://github.com/ggerganov/whisper.cpp/assets/140973261/b4693a9f-d5f5-4d85-8752-90e93ac5277a">
e020

ziegenberg commented:
Hi @Makememo,
was this a singular incident or does this happen regularly?

Makememo commented:
> Hi @Makememo,
> 
> was this a singular incident or does this happen regularly?

I tested three videos and I had this problem. 

The common feature of these videos is that there is a music section that begins to mess up the timeline.

ronyfadel commented:
@ziegenberg @bobqianic just wanted to check in on this: is it still needed? Will it land any time soon?

itsthisjustin commented:
Will test on our end now too. Anything "special" that needs to be done in order to test this using the Swift package? @ziegenberg ?

ziegenberg commented:
I have no experience with Swift, sorry.

ziegenberg commented:
Hi @bobqianic, would you be willing to rebase this once more?

ziegenberg commented:
Hi @ggerganov,

Is this being merged? Or do the changes now differ too much for this PR to be of any use?

Greetings,
Daniel
